source/extensions/filters/network/redis_proxy/command_splitter_impl.cc

```diff
@@ -101,12 +103,56 @@ void SingleServerRequest::cancel() {
   handle_ = nullptr;
 }
 
+const std::string ErrorFaultRequest::FAULT_INJECTED_ERROR = "Fault Injected: Error";
+
+SplitRequestPtr ErrorFaultRequest::create(Router&, Common::Redis::RespValuePtr&&,
+                                          SplitCallbacks& callbacks, CommandStats& command_stats,
+                                          TimeSource& time_source, bool delay_command_latency) {
+  std::unique_ptr<ErrorFaultRequest> request_ptr{
+      new ErrorFaultRequest(callbacks, command_stats, time_source, delay_command_latency)};
+
+  request_ptr->onResponse(Common::Redis::Utility::makeError(FAULT_INJECTED_ERROR));
+  command_stats.error_.inc();
+  command_stats.error_fault_.inc();
+  return request_ptr;
+}
+
+std::unique_ptr<DelayFaultRequest> DelayFaultRequest::create(SplitCallbacks& callbacks,
+                                                             CommandStats& command_stats,
+                                                             TimeSource& time_source,
+                                                             Event::Dispatcher& dispatcher,
+                                                             std::chrono::milliseconds delay) {
+  std::unique_ptr<DelayFaultRequest> delay_fault_ptr{
+      new DelayFaultRequest(callbacks, command_stats, time_source, dispatcher, delay)};
+  return delay_fault_ptr;
+}
+
+void DelayFaultRequest::onResponse(Common::Redis::RespValuePtr&& response) {
+  response_ = std::move(response);
+  dispatcher_.post([this]() {
+    delay_timer_ = dispatcher_.createTimer([this]() -> void { onDelayResponse(); });
+    delay_timer_->enableTimer(delay_);
+  });
```

Comment:
I'm not exactly sure what is happening here, but this is not right.
1) We can't use post as post can't be cancelled. You have to handle (and test) disconnection during the fault delay.
2) I'm not sure why you are posting and then creating a timer. Just create a timer?

Reply:
1. The post is needed because of (2). Once the timer is created on the correct thread, the function `void DelayFaultRequest::cancel()` on line 147 is called if we abort the request, and the timer is cancelled.
2. The post is used to get things to the correct thread. If I comment out lines 132 and 135, I get the following error on timer creation:

```
[2020-06-02 17:10:58.336][24921][info][main] [source/server/server.cc:622] starting main dispatch loop
[2020-06-02 17:11:22.949][24937][critical][assert] [source/common/event/dispatcher_impl.cc:158] assert failure: isThreadSafe().
```
Should I be creating the timer in a different way?

Reply:
There should be no cross thread behavior here. Why are we crossing threads for the delay? There should just be a timer with the delay that you can cancel?

Reply:
Yes you are using the wrong dispatcher. You need to use the worker/TLS provided dispatcher. Depending on the context of where you are running you can likely get this from the connection you are working with if the filter, or if it's in TLS context there is a dispatcher provided in the set call. This will be more obvious if you write an integration test as I suggested.

/wait

Reply:
Thanks for confirming! I've fixed that and the delay fault integration test works properly now (with no post; it failed without the post before using the correct dispatcher).